### PR TITLE
Added Mips related compoennts, change wire ui, new UI helpers, new misc compoents

### PR DESCRIPTION
**Pull request by Spooky-Firefox**
# Syncrim

## simulator common etc
- `syncrim::common::Input` now derives partial eq and eq.
- The EguiCompoennt trait was expanded with a new function `get_input_location` which is used to get a location of a port based upon its relevant input struct.
- `component_store` now saves JSON with pretty formatting.
- Changed some unwrap in editor to expect with better error messages.
- Editor now uses Wire::new
- Gui now uses primary mouse button to pan
- Added helper functions that constructs most of an ui for a component
- Added an auto wire function that uses get_input_location to create wires
- Better Error logging in Simulator
### Component changes
Most constants for out id has been changed from `"out"` to `"{component_type}_out"`.
Most components now implement `get_input_location`.

### Add 
Now supports a scale variable allowing the scale of the rendered component to be modified.

### Mem
Documented some code.

### Mux
Now supports a scale variable allowing the scale of the rendered component to be modified.

### Wire
Wire color was added, this changed how active is implemented, if component now read from is inactive wire is dashed
Changed om hover is handed, using distance from line.
In editor, added duplicate point button and color editor

### Probe edit
if unable to format as signal signed try formatting as signal unsigned

### And
bitwise logical and.

### clk
adds input with a 4.

### equals
sends a 1 if the inputs are equal 0 else (if a == b => 1).

### shift_left_const
shifts left by a constant value.

### zero_extend
zero extends.

# Mips-Lib
components required to implement a single cycle and pile lined mips

### PhysMem
This component has no input nor outputs, this is because this is basically a wrapper for a MipsMem struct which takes care of reset and unclock. This was necessary since there are multiple components that need to access the memory (IM, DM).
access to mips_mem is done in this *not so nice way*
```rust
let mut mips_mem_mut_ref = {
    let v = &simulator.ordered_components;
    let comp = v
        .into_iter()
        .find(|x| x.get_id_ports().0 == self.phys_mem_id)
        .expect(&format!("cant find {} in simulator", self.phys_mem_id));
    // deref to get Rc
    // deref again to get &dyn EguiComponent
    let comp_any = (**comp).as_any();
    let phys_mem: &PhysicalMem = comp_any
        .downcast_ref()
        .expect("can't downcast to physical memory");
    phys_mem
        .mem
        .borrow_mut()
    };
```
This should probably be updated to not panic and instead return condition::error

### MipsMem
A struct which  contains the memory as well as the symbols and sections of that memory it has methods to read and write to addresses with sizes of byte half and word, endianes can be specified. There exist version which would return error when a write is unaligned. Each write also returns a MemWriteReturn which is used with the revert method to undo an operation, note the order of which reverts are called matters, revert the newest operation first.

MipsMem also has the ability be created from an elf, note this method panics when an the passed byte slice is not en elf

### IM
The instruction memory components reads the given address from its associated physical memory, this component will panic if its physical memory is not found.
When IM encounters a break point from its MemViewWindow a halt condition is returned stopping the simulation.
There is also a button that loads an elf to memory by passing the bytes to physmem.

### DM
When the control signal is one of the read ones (read-byte, -half, -word), the data at the address is returned. If the address is unaligned the component will return an condition::error, this is also true for the write operations. When a read or write happens at an address that is a breakpoint (determined according to MemViewWindow) a condition::halt is returned.

### MemViewWindow
This is used to display a window showing the content of the passed mips_mem.
features
- Formatting options such as hex, signed decimal, utf-8 and mips assembly with symbol resolution
- Jumping to symbols and sections
- Jumping to registers, such as were sp and ra points
- Displaying sections and symbols
- Displaying where registers point
- Setting and removing breakpoints
image
image
image

### RegFile
This component have 32 internal registers that can be written to and can output 2 two of those registers. This component displays the internal registers when rendered. Format and reg names can be chosen.

### ALU
This an simple ALU which support basic orations such as add, add, sub and bit-wise operations (including shifting but not roll), other supported operations are slt sltu and lui. The operations are supplied trough the op input.

### Branch Logic
This component compares both rs and rt in in accordance to what operation is done (jr, beq and ect) and outputs a signal to control mux which determines next pc.

### Jump merger
Combines the instruction with current pc to calculate the jump address of an jump instruction.

### Instructions splitter
Takes an 32 bit instruction and splits it into rs, rt, rd, shamt, funct, immediate, and target.

### SignZeroExtend
compoents sign or zero extend the input signal according to its control input

### ControlUnit
Sends control signal to muxes, alu and other mips components according to the given instruction

## Models
This includes PR includes a functioning single cycle mips and a work in progresses pipeline version

### Single cycle MIPS
image

### Pipe-lined MIPS
Work In Progress

### CODE REVIEW

File: src/components/clk.rs
Comment (onsdagens) [inline src/components/clk.rs:1]: can this be removed?

File: src/autowire.rs
Comment (onsdagens) [inline src/autowire.rs:1]: The warnings here can be allowed

File: mips-lib/src/main.rs [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+use std::path::PathBuf;
+#[cfg(feature = "gui-egui")]
+use syncrim::gui_egui::editor::Library;
+use syncrim::{common::ComponentStore, fern::fern_setup};
+/// Simple program to greet a person
+#[derive(Parser, Debug)]
+#[command(author, version, about, long_about = None)]
+struct Args {
+    /// Path to the model to load on startup
+    #[arg(short, long, default_value = "empty.json")]
```
Comment (onsdagens) [inline mips-lib/src/main.rs:12]: maybe change this to the single cycle mips?

File: mips-lib/src/components/mips_jump_merge.rs [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    pub(crate) instr_addr_in: Input,
+    pub(crate) jump_addr_in: Input,
+}
+
+#[typetag::serde]
+impl Component for JumpMerge {
+    fn to_(&self) {
+        trace!("jump_merge");
+    }
+    // #[cfg(feature = "gui-egui")]
```
Comment (onsdagens) [inline mips-lib/src/components/mips_jump_merge.rs:28]: dead code can be removed

File: src/gui_egui/editor.rs [hunk 3]
```diff
@@ -127,12 +128,18 @@ impl Default for Library {
...
                 select: dummy_input.clone(),
                 m_in: vec![dummy_input.clone(), dummy_input.clone()],
+                scale: 1.0,
             }),
             Rc::new(Register {
                 id: "reg".to_string(),
                 pos: (0.0, 0.0),
                 r_in: dummy_input.clone(),
             }),
+            Rc::new(MIPSCLK {
```
Comment (onsdagens) [inline src/gui_egui/editor.rs:138]: is this needed?

File: src/gui_egui/editor.rs [hunk 4]
```diff
@@ -357,7 +364,10 @@ impl Editor {
                 _ => e.components.retain_mut(|c| {
                     let old_key = c.as_ref().get_id_ports().0;
                     let mut context = e.contexts.remove(&old_key).unwrap();
-                    let render_return = (*Rc::get_mut(c).unwrap()).render_editor(
+                    let debug_id = c.get_id_ports().0;
+                    let render_return = (*Rc::get_mut(c)
+                        .expect(&format!("More than one reference exist to {}, can't get mut, therefore not render editor", debug_id)))
```
Comment (onsdagens) [inline src/gui_egui/editor.rs:369]: just #[allow] this lgtm

**Review by onsdagens — CHANGES_REQUESTED** [review]
Couple of comments

File: mips-lib/src/components/mips_jump_merge.rs (hunk 2 above)
Comment (onsdagens) [reply mips-lib/src/components/mips_jump_merge.rs:28]: there is a bunch of random commented code scattered about, we can probably get rid of it.

File: src/autowire.rs
Comment (Spooky-Firefox) [reply src/autowire.rs:1]: fixed in 3c77dbf7ddc4bc5016e9541505ee179077ecf897

File: src/gui_egui/editor.rs (hunk 4 above)
Comment (Spooky-Firefox) [reply src/gui_egui/editor.rs:370]: fixed in 3c77dbf7ddc4bc5016e9541505ee179077ecf897

File: mips-lib/src/main.rs (hunk 1 above)
Comment (Spooky-Firefox) [reply mips-lib/src/main.rs]: fixed in d19ff005c6cada6c3a12b9012063ec30e0cd3c5c

File: src/gui_egui/editor.rs (hunk 3 above)
Comment (Spooky-Firefox) [reply src/gui_egui/editor.rs]: fixed in 1fadfac92d277c0ca8defc5a5deaf9a1c460c578

File: src/components/clk.rs
Comment (Spooky-Firefox) [reply src/components/clk.rs:1]: fixed 1fadfac92d277c0ca8defc5a5deaf9a1c460c578

**Review by onsdagens — APPROVED** [review]
LGTM

**Pull request merged**